### PR TITLE
Int32 type support

### PR DESCRIPTION
**Pull request by shmsong**
This PR introduces Int32 as a data type, changes including

- Support for Int32 as TensorView data type
- May add Int32 as IR node so Int32 scalars can be used if needed (currently removed from this PR).
- Refactored python and cpp test cases for better integer test support

Addtional changes:
- Add type promotion and demotion check for backward graph
- Handled binary ops corner cases on integers:
- Current catches: (may end up not fusing some of the below ops applied to all_integer inputs)
     - Difficult ops:
         - torch.div:
             eager mode torch.div actually does floating point division, while `BinaryOpType::div` does integer division.
         - torch.pow: not natively supported for integer inputs
         - torch.atan2: unimplemented for integer inputs
         - torch.fmod: unimplemented for integer inputs

### CODE REVIEW

File: test/test_jit_cuda_fuser.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 from torch.testing._internal.codegen.random_topo_test import runDefaultTestWithSeed
 from torch.testing import FileCheck
 
-from test_jit import JitTestCase, RUN_CUDA
+from torch.testing._internal.jit_utils import JitTestCase, RUN_CUDA
```
Comment (shmsong) [inline test/test_jit_cuda_fuser.py:11]: Note:
Importing from test_jit actually changes the behavior of eager mode for some reason. Will revisit this later.

**Comment by tlemo** [comment]
> * Added Int32 as IR node so Int32 scalars can be used (likely won't need, may remove this before merging)

Wouldn't be easier not to add this in the first place (rather than adding and removing it later)?

**Comment by shmsong** [comment]
> > * Added Int32 as IR node so Int32 scalars can be used (likely won't need, may remove this before merging)
> 
> Wouldn't be easier not to add this in the first place (rather than adding and removing it later)?

I completely agree. I should have staged the commits that way. I just removed the IR nodes part for now. Thanks for pointing this out. 👍

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp [hunk 2]
```diff
@@ -98,6 +98,9 @@ bool validateKernelArgTensor(
     case at::ScalarType::Long:
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor_utils.cpp:98]: Should we do a follow up to make our DataType consistent with theirs? Long/Int vs Int/Int32

**Review by csarofeen — APPROVED** [review]
LGTM, thanks!

**Review by csarofeen — APPROVED** [review]
LGTM. Thanks.

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp (hunk 2 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp:98]: Sure. It'd be nice. I will do that in a follow up PR.
